artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java

```diff
@@ -511,22 +534,88 @@ private void connectSender(Queue queue,
                sender.setProperties(mapProperties);
             }
 
+            if (desiredCapabilities != null) {
+               sender.setDesiredCapabilities(desiredCapabilities);
+            }
+
             AMQPOutgoingController outgoingInitializer = new AMQPOutgoingController(queue, sender, sessionContext.getSessionSPI());
 
+            sender.open();
+
             ProtonServerSenderContext senderContext = new ProtonServerSenderContext(protonRemotingConnection.getAmqpConnection(), sender, sessionContext, sessionContext.getSessionSPI(), outgoingInitializer).setBeforeDelivery(beforeDeliver);
+            // This will be done on the remote open
+            sender.setContext(new Runnable() {
+               @Override
+               public void run() {
```

Comment:
@gemmellr would I still get a remote open on such case?

Reply:
Yes, an attach frame is always sent and then followed by a detach, due to the symmetry of the protocol. For a sender, youll get a null remote target to show no terminus was created amd hint that you are about to get a detach with error explaining why. (I would have done it differently personally).

http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#doc-idp315568
Figure 2.33

Reply:
The error would then cause the exception anyway on this case. Can't we leave it as it is?


I have a Test called ValidateAMQPErrorsTest

Do you think you could add a test validating such case please?

Reply:
It might but perhaps only after considering it open and potentially trying to use it, feeding it through to the rest of the broker as a consumer on the queue etc. Would typically be nicer to treat it as a failure immediately, even if that means just not doing anything. I can take a look yep.

Reply:
@gemmellr I wouldn't know how to create a test for such condition. I would love to do this but I would need help.

Reply:
Essentially its just having the peer set the appropriate terminus to be null (in this case, the target terminus as its a producer) and then closing the link with an error when you would normally just be opening the link. I got distracted by other thigns this afternoon, ill look on monday.

Reply:
I added a commit at https://github.com/apache/activemq-artemis/commit/58b65283fdb1462c6ac01708ecfc4141dd079ea7 including a new debug test to show the issues, it manipulates a delay between the 'refusing attach' and the detach such that the broker actually fires up a consumer on the queue, and also flows credit (unusual during a refusal but possible) so it even sends messages across before it gets the detach. It should instead notice the refusal and skip doing anything. The debug test also (if the noted code is commented out) highlights the way the broker falls over the lack of caps/props due to overlooking the refusal, whereas it wouldn't if it handled the refusal (e.g set a limit on how long it will wait for the detach to explain the issue, then kill the bits if it doesnt happen in that time).

I noticed when creating it that that the test sometimes failed due to receiving more than the expected number of retries, particularly when that code was uncommented, so I think there may be some issues lurking in the reconnect stuff. Some of the other tests you have might be useful to find them if they more precisely track the total number of connections rather than just the number of disconnects that occur.

The commit also has a couple cleanup suggestions for other tests (which could apply to some of the others as well)

Reply:
@gemmellr I could not figure out what you said about commented out code... what code you were referring to?

I am changing the tests on ValidateAMQPErrorsTest to setTarget(getRemoteTarget()) and adding a check if the remote is null for this.


this is a bit beyond the original intent of this PR.. which was more towards the dual mirror addition than this kind of check.

I will add these as part of this PR, but we can improve this as part of a separate JIRA / PR. you ok with that?

Reply:
By 'the noted code' I meant: https://github.com/gemmellr/activemq-artemis/commit/58b65283fdb1462c6ac01708ecfc4141dd079ea7#diff-18d73a3fc9014f71a70754aae490f5e0765b10fd76bb44daf6a50e5cea85dbc5R461-R467

The latter bit of my comment had a mistake, where I said 'when that code was uncommented' it should have been 'when that code was commented'( i.e removed per the notes).

I think its good to add handling the null terminus in this PR at least, when its already adding a bunch of stuff in that exact space that could make it otherwise misbehave and mislead folks with unhelpful errors if it didnt....though I wouldnt really consider fully handling the basic link refusal beyond the scope of the PR given its exactly how the broker would e.g refuse it due to permissions.

Reply:
I have added a commit to handle the null remoteTarget for now...

Do you want to (or can you?) open a JIRA for further verifications after this task?